kythe/go/serving/pipeline/nodes/nodes.go

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2018 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package nodes provides Beam transformations over *ppb.Nodes.
+package nodes
+
+import (
+	"reflect"
+	"sort"
+
+	"kythe.io/kythe/go/services/graphstore/compare"
+	"kythe.io/kythe/go/util/schema"
+	"kythe.io/kythe/go/util/schema/edges"
+	"kythe.io/kythe/go/util/schema/facts"
+
+	"github.com/apache/beam/sdks/go/pkg/beam"
+
+	ppb "kythe.io/kythe/proto/pipeline_go_proto"
+	scpb "kythe.io/kythe/proto/schema_go_proto"
+	spb "kythe.io/kythe/proto/storage_go_proto"
+)
+
+func init() {
+	beam.RegisterFunction(embedSourceKey)
+	beam.RegisterFunction(entryToNode)
+
+	beam.RegisterType(reflect.TypeOf((*combineNodes)(nil)).Elem())
+}
+
+// FromEntries transforms a PCollection of *ppb.Entry protos into *ppb.Nodes.
+func FromEntries(s beam.Scope, entries beam.PCollection) beam.PCollection {
+	return beam.ParDo(s, embedSourceKey,
+		beam.CombinePerKey(s, &combineNodes{},
+			beam.ParDo(s, entryToNode, entries)))
+}
+
+func entryToNode(e *spb.Entry) (*spb.VName, *ppb.Node) {
+	n := &ppb.Node{}
+	if e.EdgeKind == "" {
```

Comment:
Does an Entry ever simultaneously encode (a) facts for a node and (b) edge data associated with that node?  This code implicitly assumes that it doesn't.  If it never should, this code should probably verify that it doesn't; if it could, then this logic needs to be structured a bit differently.

Reply:
It is the case that an Entry is either a fact or edge, but not both, based on which fields are set: (source, edge_kind, target) or (source, fact_name, fact_value).  There's some legacy around setting the fact_name to "/" for edges, but that's not required anymore.

I'm not sure what you mean by verify.  There are a couple of options in Beam: ignore, crash, or log.  Currently chosen was "ignore".  Generally, the Kythe design prefers to not crash under any circumstance and provide lossy data over no data, but this could be considered a critical error so I'll go with that for now.